### PR TITLE
Added the ability to use indexes in format arguments

### PR DESCRIPTION
**Pull request by APaloski**
In response to issue #238 I added in the ability to use indexes in the format strings.

I used the $-index-format syntax discussed in that thread and made the index start at one to be consistent with the way that java formatters act.

### CODE REVIEW

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 1]
```diff
@@ -106,38 +107,58 @@ public Builder add(String format, Object... args) {
...
+          /* Rules:
+           * Format must be in the form $[Numeric Index]{N, L, S, T}
+           * OR
+           * in the form ${$, >, <, [, ]}
+           * Numeric index is ONE BASED (for consistency with java formatting)
+           * Numeric indexing must refer to a format argument within the args array
+           * Only progress iteration over the argument iterator if we do not hit an indexed entry
+           */
+          int countOfIndexCharacters = 0;
+          while (Character.isDigit(format.charAt(p + countOfIndexCharacters + 1))) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: We probably want a much simpler, much dumber set of index characters. `c >= '0' && c <= '9'`. Character.isDigit supports lots of "digit" like characters!

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 2]
```diff
@@ -106,38 +107,58 @@ public Builder add(String format, Object... args) {
...
+            countOfIndexCharacters++;
+          }
+          int argsIndex = -1;
+          if (countOfIndexCharacters != 0) {
+            argsIndex = Integer.parseInt(format.substring(p + 1, p + 1 + countOfIndexCharacters));
+            checkArgument(!(argsIndex > args.length),
+                "Argument index %s in '%s' is larger than number of parameters",
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: Awesome

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 3]
```diff
@@ -106,38 +107,58 @@ public Builder add(String format, Object... args) {
...
+           * Only progress iteration over the argument iterator if we do not hit an indexed entry
+           */
+          int countOfIndexCharacters = 0;
+          while (Character.isDigit(format.charAt(p + countOfIndexCharacters + 1))) {
+            countOfIndexCharacters++;
+          }
+          int argsIndex = -1;
+          if (countOfIndexCharacters != 0) {
+            argsIndex = Integer.parseInt(format.substring(p + 1, p + 1 + countOfIndexCharacters));
+            checkArgument(!(argsIndex > args.length),
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: Maybe simplify to `checkArgument(argsIndex < args.length, ...`

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 4]
```diff
@@ -106,38 +107,58 @@ public Builder add(String format, Object... args) {
...
             case 'T':
-              this.args.add(argToType(i.next()));
+              this.args.add(argToType((argsIndex == -1) ? i.next() : args[argsIndex - 1]));
               break;
             case '$':
             case '>':
             case '<':
             case '[':
             case ']':
+              checkState(argsIndex == -1, "$$, $>, $<, $[ and $] may not have an index");
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: Sweet.

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 5]
```diff
@@ -106,38 +107,58 @@ public Builder add(String format, Object... args) {
         if (format.charAt(p) != '$') {
           nextP = format.indexOf('$', p + 1);
           if (nextP == -1) nextP = format.length();
+          formatParts.add(format.substring(p, nextP));
         } else {
           checkState(p + 1 < format.length(), "dangling $ in format string %s", format);
-          switch (format.charAt(p + 1)) {
+          /* Rules:
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: Nice doc.

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 6]
```diff
@@ -106,38 +107,58 @@ public Builder add(String format, Object... args) {
...
             case ']':
+              checkState(argsIndex == -1, "$$, $>, $<, $[ and $] may not have an index");
               break;
             default:
               throw new IllegalArgumentException("invalid format string: " + format);
           }
 
-          nextP = p + 2;
+          nextP = p + countOfIndexCharacters + 2;
+          formatParts.add(countOfIndexCharacters == 0
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: Does it do the right thing automatically without the ternary? I think it might just work

File: src/test/java/com/squareup/javapoet/CodeBlockTest.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
+package com.squareup.javapoet;
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/CodeBlockTest.java]: please copy & paste a square copyright header from another file!

File: src/test/java/com/squareup/javapoet/CodeBlockTest.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    try {
+      CodeBlock.builder().add("$T $2T", String.class).build();
+      fail();
+    } catch (IllegalArgumentException exp) {
+      assertThat(exp).hasMessage("Argument index 2 in '$T $2T' is larger than number of parameters");
+    }
+  }
+  
+  @Test
+  public void indexTooLow() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/CodeBlockTest.java]: what about `0` ?

File: src/test/java/com/squareup/javapoet/CodeBlockTest.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+      fail();
+    } catch (IllegalArgumentException exp) {
+      assertThat(exp).hasMessage("invalid format string: $T $-1T");
+    }
+  }
+  
+  @Test
+  public void indexingDoesNotIncreaseNaturalIndex() {
+    CodeBlock block = CodeBlock.builder().add("$L $L $2L $L", 1, 2, 3).build();
+    assertThat(block.toString()).isEqualTo("1 2 2 3");
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/CodeBlockTest.java]: great test

File: src/test/java/com/squareup/javapoet/CodeBlockTest.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+  }
+  
+  @Test
+  public void indexingCanBeInterleved() {
+    CodeBlock block = CodeBlock.builder().add("$L $3L $L $2L $L $1L", 1, 2, 3).build();
+    assertThat(block.toString()).isEqualTo("1 3 2 2 3 1");
+  }
+  
+  @Test
+  public void sameIndexCanBeUsedWithDifferentFormats() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/CodeBlockTest.java]: neato

File: src/test/java/com/squareup/javapoet/TypeSpecTest.java [hunk 11]
```diff
@@ -1043,6 +1043,32 @@ public void interfaceDefaultMethods() throws Exception {
         + "  }\n"
         + "}\n");
   }
+  
+  @Test public void indexedElseIf() throws Exception {
+    TypeSpec taco = TypeSpec.classBuilder("Taco")
+        .addMethod(MethodSpec.methodBuilder("choices")
+            .beginControlFlow("if ($1L != null || $1L == $2L)", "mTaco", "mOtherTaco")
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/TypeSpecTest.java]: mind dropping the `m` prefix here? We don't want to endorse that code style, not even in generated code!

File: src/test/java/com/squareup/javapoet/TypeSpecTest.java [hunk 12]
```diff
@@ -1043,6 +1043,32 @@ public void interfaceDefaultMethods() throws Exception {
...
+        .build();
+    assertThat(toString(taco)).isEqualTo(""
+        + "package com.squareup.tacos;\n"
+        + "\n"
+        + "import java.lang.System;\n"
+        + "\n"
+        + "class Taco {\n"
+        + "  void choices() {\n"
+        + "    if (mTaco != null || mTaco == mOtherTaco) {\n"
+        + "      System.out.println(\"only one taco? NOO!\");\n"
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/TypeSpecTest.java]: This is hilarious.

File: src/test/java/com/squareup/javapoet/TypeSpecTest.java [hunk 13]
```diff
@@ -1818,15 +1844,6 @@ private String toString(TypeSpec typeSpec) {
     }
   }
 
-  @Test public void tooManyArguments() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/TypeSpecTest.java]: Interesting. Yup, that makes sense.

**Comment by swankjesse** [comment]
This is wonderful, thank you!

Please complete our contributors paperwork so we can redistribute your code:
https://github.com/square/javapoet/blob/master/CONTRIBUTING.md

**Comment by APaloski** [comment]
I added in fixes to the comments you left above and completed the contributors form. I hope all looks well.

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 14]
```diff
@@ -106,41 +107,69 @@ public Builder add(String format, Object... args) {
         if (format.charAt(p) != '$') {
           nextP = format.indexOf('$', p + 1);
           if (nextP == -1) nextP = format.length();
+          formatParts.add(format.substring(p, nextP));
         } else {
-          checkState(p + 1 < format.length(), "dangling $ in format string %s", format);
-          switch (format.charAt(p + 1)) {
+          checkState(p <= format.length(), "dangling $ in format string %s", format);
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: why this change ? I think this is a bug

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 15]
```diff
@@ -106,41 +107,69 @@ public Builder add(String format, Object... args) {
...
+          /* Rules:
+           * Format must be in the form $[Numeric Index]{N, L, S, T}
+           * OR
+           * in the form ${$, >, <, [, ]}
+           * Numeric index is ONE BASED (for consistency with java formatting)
+           * Numeric indexing must refer to a format argument within the args array
+           * Only progress iteration over the argument iterator if we do not hit an indexed entry
+           */
+          int countOfIndexCharacters = 0;
+          while (isSimpleDigit(format.charAt(p + countOfIndexCharacters + 1))) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: does this crash on malformed format strings like "$" ?

File: src/main/java/com/squareup/javapoet/CodeBlock.java [hunk 16]
```diff
@@ -106,41 +107,69 @@ public Builder add(String format, Object... args) {
...
+            countOfIndexCharacters++;
+          }
+          int argsIndex = -1;
+          if (countOfIndexCharacters != 0) {
+            argsIndex = Integer.parseInt(format.substring(p + 1, p + 1 + countOfIndexCharacters));
+            checkArgument(argsIndex <= args.length,
+                "Argument index %s in '%s' is larger than number of parameters",
+                argsIndex, format);
+            checkArgument(argsIndex > 0,
```
Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: I think you can drop this case; we never accept the `-` character anywhere.

Comment (swankjesse) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: ... nope I'm wrong. `0`.

File: src/test/java/com/squareup/javapoet/CodeBlockTest.java [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+
+import org.junit.Test;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.fail;
+
+public class CodeBlockTest {
+  
+  @Test
+  public void indentCannotBeIndexed() {
```
Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/CodeBlockTest.java]: Please add a test that uses "$" all by itself, and another one that tests "$1" ? I think both of these will crash in the wrong way

Comment (swankjesse) [inline src/test/java/com/squareup/javapoet/CodeBlockTest.java]: (they should throw exceptions with nice messages)

**Comment by swankjesse** [comment]
I checked your CLA. Looks good.

Mind adding two more short test cases that might catch a problem in the code? Also please use http://rebaseandsqua.sh to get only a single commit for this change.

And then I'll merge this. Thanks!

File: src/main/java/com/squareup/javapoet/CodeBlock.java (hunk 14 above)
Comment (APaloski) [inline src/main/java/com/squareup/javapoet/CodeBlock.java]: My mistake, changed the wrong line based upon previous comments and didn't catch it. Fixed in next check in.

**Comment by svenjacobs** [comment]
Just when I was about to get started on #238 I see that the work has already been done. That's great :+1: :) Looking forward for this being merged.

**Comment by swankjesse** [comment]
My apologies for sitting on this for 17 days. I missed the notification on your updated push.

LGTM, thanks!
